easybuild/easyblocks/q/quantumespresso.py

```diff
@@ -27,232 +27,415 @@
 
 @author: Kenneth Hoste (Ghent University)
 @author: Ake Sandgren (HPC2N, Umea University)
+@author: Davide Grassano (CECAM, EPFL)
 """
 import fileinput
 import os
 import re
 import shutil
 import sys
-from easybuild.tools import LooseVersion
 
 import easybuild.tools.environment as env
 import easybuild.tools.toolchain as toolchain
-from easybuild.easyblocks.generic.configuremake import ConfigureMake
 from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools import LooseVersion
 from easybuild.tools.build_log import EasyBuildError
 from easybuild.tools.filetools import copy_dir, copy_file
 from easybuild.tools.modules import get_software_root, get_software_version
+from easybuild.tools.run import run_cmd
+
+from easybuild.easyblocks.generic.configuremake import ConfigureMake
 
 
 class EB_QuantumESPRESSO(ConfigureMake):
     """Support for building and installing Quantum ESPRESSO."""
 
+    TEST_SUITE_DIR = "test-suite"
+
     @staticmethod
     def extra_options():
         """Custom easyconfig parameters for Quantum ESPRESSO."""
         extra_vars = {
             'hybrid': [False, "Enable hybrid build (with OpenMP)", CUSTOM],
             'with_scalapack': [True, "Enable ScaLAPACK support", CUSTOM],
             'with_ace': [False, "Enable Adaptively Compressed Exchange support", CUSTOM],
+            'with_fox': [False, "Enable FoX support", CUSTOM],
+            'with_epw': [True, "Enable EPW support", CUSTOM],
+            'with_gipaw': [True, "Enable GIPAW support", CUSTOM],
+            'with_wannier90': [False, "Enable Wannier90 support", CUSTOM],
+            'test_suite_targets': [[
+                "pw", "pp", "ph", "cp", "hp", "tddfpt", "epw",
+                ], "List of test suite targets to run", CUSTOM],
+            'test_suite_allow_failures': [[
+                'relax',  # Too strict thresholds
+                'epw_polar',  # Too strict thresholds
+                'cp_h2o_scan_libxc',  # Too strict thresholds
+                'hp_metal_us_magn',  # Too strict thresholds
+                'hp_soc_UV_paw_magn',  # In 7.3 test has more params than the baseline
+                'ph_ahc_diam',  # Test detects a ! as an energy in baseline
+                'tddfpt_magnons_fe',  # Too strict thresholds
+            ], "List of test suite targets that are allowed to fail (name can partially match)", CUSTOM],
+            'test_suite_threshold': [0.97, "Threshold for test suite success rate", CUSTOM],
```

Comment:
With the falkyness of some tests failing just because the absolute/relative errors are slightly higher than the thresholds sets in the baselines, i think this could be tricky.
What I am mostly worried about is that if that number is not carefully curated we might be missing some segfaults that could arise.
This is why before i added what should be the flaky test to `test_suite_allow_failures` and raised (removed in commit [f182aea](https://github.com/easybuilders/easybuild-easyblocks/pull/3241/commits/f182aea69378c31e8e9b5099b54142ae827b3709)) if a test not in that list failed (most likely it would fail because the calculation does not actually finish)

Reply:
> With all the testing I've done, I have the exact numbers for almost everything already, I just need to dig them out. We can then add them to the easyconfigs and do a final rerun of the builds. If we do a rerun of builds, do you think we should add the `fast-math` as well to see how we do?

Sure we could try. This night i just finished testing an easyconfig for 7.3 with the option. I think it should work for all versions just by adding

```
    'extra_cflags': '-ffast-math',
    'extra_fflags': '-ffast-math',
    'extra_fcflags': '-ffast-math',
    'extra_f90flags': '-ffast-math',
```

to the toolchain options

Reply:
> With the falkyness of some tests failing just because the absolute/relative errors are slightly higher than the thresholds sets in the baselines, i think this could be tricky. What I am mostly worried about is that if that number is not carefully curated we might be missing some segfaults that could arise. This is why before i added what should be the flaky test to `test_suite_allow_failures` and raised (removed in commit [f182aea](https://github.com/easybuilders/easybuild-easyblocks/pull/3241/commits/f182aea69378c31e8e9b5099b54142ae827b3709)) if a test not in that list failed (most likely it would fail because the calculation does not actually finish)

But i guess we can use the `failures` array to check the number of failures without the ignored ones.
I would still leave the threshold on the total number (without the ignored ones) though as the ignored tests including `relax` actually could be excluding a non trivial number of tests failures